Add the skeleton of an AWS API Gateway + Lambda frontend service

wjbuys opened the pull request:
#### Summary
This adds a Spring MVC/REST web service, running as a Lambda function, fronted by API Gateway. See `docs/frontend.md` for an overview of how this works.

Since this is the first real commit to the repository it also sets up [`spotless`](https://github.com/diffplug/spotless/tree/master/plugin-gradle) for enforcing and applying Java code style according to the [Google Java Styleguide](https://google.github.io/styleguide/javaguide.html). This choice was driven mainly by the need for a style that is easy to automatically apply through build scripts and IDE plugins. See `CONTRIBUTION.md` for details.

#### Implementation details
The main moving parts are:
- Spring controller from which Swagger model is derived: `com.amazonaws.blox.frontend.controllers.EnvironmentController`
- Serverless Application Model template: `frontend-infrastructure/api/template.yml`
- Generated Swagger model: `frontend-service/api/swagger.yml`

There is some custom build logic in `buildSrc` that takes care of generating a Swagger model from the Spring controller, and to add all the necessary extensions to make it work with API Gateway.

I've separated the auto-generated client into a separate commit to make it clear which code is auto-generated for ease of reviewing.

#### Testing
The change includes some basic unit and end-to-end tests, mostly just to test the happy path for integration.

#### Licensing

This contribution is under the terms of the Apache 2.0 License: Yes

Commit:
Add the skeleton of an AWS API Gateway + Lambda frontend service

This adds a Spring MVC/REST web service, running as a Lambda function,
fronted by API Gateway. See `docs/frontend.md` for an overview of how
this works

The main moving parts are:
- Spring controller from which Swagger model is derived: `com.amazonaws.blox.frontend.controllers.EnvironmentController`
- Serverless Application Model template: `frontend-infrastructure/api/template.yml`
- Generated Swagger model: `frontend-service/api/swagger.yml`

There is some custom build logic in `buildSrc` that takes care of
generating a Swagger model from the Spring controller, and to add all
the necessary extensions to make it work with API Gateway.

This commit does not include the client that gets generated from the
API, for ease of reviewing.

Commit:
Check in the generated client for the Frontend API

This code was generated by running:

```
./gradlew updateClient
```

arminc commented on docs/design.md:
```diff
@@ -0,0 +1,33 @@
+* It improves active container reuse. Basically, due to the way lambda routing works, if you have all requests going to a single lambda function, each request is more likely to hit an active container, resulting in lower latency.
+
+### Why use a Servlet implementation in the lambda function?
+Instead of using `aws-serverless-java-container` and using a Servlet implementation, we could some of our own glue code to correctly call the right code in response to incoming API calls. However, using a Servlet implementation that supports writing RESTful controllers has some compelling benefits:
```
Sentence wrong? "we could some of our own glue code"  -> "we could use/create some of our own glue code"

wjbuys replied:
Looks like copy paste mistake, should be *"we could **write** some of our own glue code"*.

wjbuys replied:
Fixed in latest revision, also moved the file to `frontend_design.md` to make the distinction clear.

arminc reviewed (commented):
Small documentation mistake

narehayrapetyan commented on CONTRIBUTING.md:
```diff
@@ -7,6 +7,9 @@ If you would like to make a significant change, it's a good idea to first open a
 Development takes place against the dev branch of this repository and pull requests should be opened against that branch.
 
+### Code Style
+This project follows the [Google Java Styleguide](https://google.github.io/styleguide/javaguide.html), and this style is enforced as part of the `check` task. We recommend you install [the `google-java-format` plugin for your IDE](https://github.com/google/google-java-format), or use the `gradle spotlessApply` task to format code before checking in.
```
is this or should it be applied as part of build?

wjbuys replied:
The `spotless` plugin will verify that all files meet the configured style, but it will not modify source files without explicitly being requested through the `spotlessApply` task. I think it should not modify source as part of a normal build, to avoid problems with a CI build potentially building code that is different from what was checked in.

narehayrapetyan replied:
How do we ensure that the style has been applied? Or are you saying that the check will run as part of the build and fail if the style is not preserved but it won't apply fixes?

wjbuys replied:
Correct.

narehayrapetyan commented on README.md:
````diff
@@ -9,6 +9,66 @@ Blox is being delivered as a managed service via the Amazon ECS Console, API and
+To run the full unit test suite, run:
+
+```
+./gradlew check
````
why check instead of unit-tests?

wjbuys replied:
That's the Gradle idiom, and the default task that is provided by the Java plugin: https://docs.gradle.org/3.3/userguide/java_plugin.html#sec:java_tasks

I figured it was easier to exclude the end-to-end tests from the default task run by CI (see the `build.gradle` in the `end-to-end-tests` project), instead of reconfiguring every single unit test task to have a different name.

narehayrapetyan commented on README.md:
````diff
@@ -9,6 +9,66 @@ Blox is being delivered as a managed service via the Amazon ECS Console, API and
+    ```
+
+  These permissions are pretty broad, so we recommend you use a separate, test account.
+- configure the `default` profile with the AWS credentials for the user you created above
````
I'm not sure if we should recommend that they change their default user to this, it really depends on what the customer is doing most often.

wjbuys replied:
Good point. What do you suggest instead? We have to default to some profile name when not provided...

narehayrapetyan replied:
As we talked offline, I don't think we should rely on profile name at all since that is unrelated to any resources and also can change without affecting resources.

wjbuys replied:
Sure, and as mentioned offline, I'll fix in a followup CR for #228

narehayrapetyan commented on build.gradle:
```diff
@@ -1,3 +1,103 @@
+    println("Profile: ${awsProfile}")
+
+    awsCli = "/usr/local/bin/aws"
+    awsRegion = "us-west-2"
```
should this be configurable?

wjbuys replied:
Perhaps. I'm not sure how yet, though. Maybe wait until we see how we may want to reconfigure this?

(Gradle's only configuration mechanism is pretty cumbersome, you have to include `-pKey=Value` on the command line, there doesn't seem to be an easy way to set persistent local configuration).

wjbuys replied:
I take that back, I totally missed [the section on `gradle.properties`](https://docs.gradle.org/current/userguide/build_environment.html#sec:gradle_properties_and_system_properties) in the docs. It looks like it should be pretty easy to put any necessary config there. As mentioned in comment below, I'll do that in a followup CR.

narehayrapetyan commented on build.gradle:
```diff
@@ -1,3 +1,103 @@
+    awsPrefix = System.getenv("USER")
+
+    stackName = "${awsPrefix}-blox-frontend-${awsRegion}"
+    s3BucketName = "${stackName}-${awsProfile}"
```
should we be tying the name to the local profile name? what if you move your credentials into a different profile or create a different user?

wjbuys replied:
Open to suggestions; the intent behind this was for my own purposes of testing stuff in my personal AWS account and the blox beta account, and this was the easiest way to namespace things at the time. Unless I'm missing something, I don't think we can actually programmatically get at a more fine-grained value to namespace on, so that would have to be yet another configuration setting.

narehayrapetyan commented on build.gradle:
```diff
@@ -1,3 +1,103 @@
+
+    stackName = "${awsPrefix}-blox-frontend-${awsRegion}"
+    s3BucketName = "${stackName}-${awsProfile}"
+    stageName = "Beta"
```
configurable?

wjbuys replied:
Can we quickly whiteboard out how we want to make this configurable? As mentioned in another comment, all this AWS settings are just super draft to make it work.

narehayrapetyan commented on buildSrc/build.gradle:
```diff
@@ -1,3 +1,20 @@
+    compile gradleApi()
+    compile localGroovy()
+
+    compile 'com.github.kongchen:swagger-maven-plugin:+'
```
is this dependency going to be included in all artifacts? We don't need swagger everywhere do we?

wjbuys replied:
No, as indicated by its location in `buildSrc`, it's a compile-dependency of our build logic. It will never be included in any artifacts.

narehayrapetyan commented on buildSrc/src/main/java/com/amazonaws/blox/swagger/ApiGatewayExtensionsFilter.java:
```diff
@@ -0,0 +1,80 @@
+  @Input private String lambdaFunctionArnTemplate;
+
+  public ApiGatewayExtensionsFilter(String lambdaFunctionArnTemplate) {
+    this.lambdaFunctionArnTemplate = lambdaFunctionArnTemplate;
```
@allargsconstructor?

wjbuys replied:
Oh right, missed this, will update.

narehayrapetyan commented on buildSrc/src/main/java/com/amazonaws/blox/swagger/ApiGatewayExtensionsFilter.java:
```diff
@@ -0,0 +1,80 @@
+  }
+
+  public Map<String, Object> defaultExtensions() {
+    HashMap<String, Object> extensions = new HashMap<String, Object>();
```
HashMap -> Map?

wjbuys replied:
Whoops...

narehayrapetyan commented on buildSrc/src/main/java/com/amazonaws/blox/tasks/PostProcessCloudformation.java:
```diff
@@ -0,0 +1,116 @@
+   * The name of the API GW resource in the SAM template into which the swagger template should be
+   * embedded.
+   */
+  @Input private String apiName;
```
Would it be useful to have an option of passing these in from command line or another script instead of only accepting input from gradle? (Maybe something that can be done later)

wjbuys replied:
I think later. Since it's tightly tied to the names checked in in the template file itself, I don't think it's particularly useful to override it from the command line rather than source code.

narehayrapetyan commented on buildSrc/src/main/java/com/amazonaws/blox/swagger/ApiGatewayExtensionsFilter.java:
```diff
@@ -0,0 +1,80 @@
+ * <p>This will add the necessary x-amazon-apigateway-integration section to all operations in a
+ * Swagger spec to correctly proxy all requests to a single lambda function.
+ */
+public class ApiGatewayExtensionsFilter implements SwaggerFilter {
```
Do we need these classes for all projects or just frontend?

wjbuys replied:
We only need them for the frontend, but they do no harm in the other projects. The reason that these are shared, is that Gradle is a bit inflexible in how it supports code that should be available in the build script itself. According to [the docs](https://docs.gradle.org/current/userguide/custom_tasks.html#sec:packaging_a_task_class), the only options are:
1. embed it in the script itself
2. put it in the `buildSrc` subproject, of which there can only be one per root-project
3. put it in a completely separate project, and consume it through a Maven repo

I think once we're comfortable that we are done iterating on this logic, we can move these from (2) to (3). However, since (3) implies making a separate repository and pushing to a Maven repo, I'd like to punt on it for now. I'll add a TODO to the code to that effect?

narehayrapetyan replied:
I was mostly concerned about code size because we have to keep the jar as small as possible for Lamdba. What do you think? I think a separate Maven repo is overkill, however.

narehayrapetyan commented on docs/frontend_design.md:
```diff
@@ -0,0 +1,33 @@
+
+* Our Cloudformation template and Swagger model are simpler. Since all operations are wired up to a single lambda function, there's only one value to wire through to the Swagger model.
+* It's easier to reason about deployments. Since we only have two AWS resources (the API and the Lambda function) we can easily version them for more deterministic deployment.
+* It improves active container reuse. Basically, due to the way lambda routing works, if you have all requests going to a single lambda function, each request is more likely to hit an active container, resulting in lower latency.
```
It's not because of routing is it? It's because containers are shut down if they don't get requests and more code in the same container means it will be hit more frequently.

wjbuys replied:
I'll update this to reflect that, and refer to the appropriate public docs.

narehayrapetyan commented on frontend-infrastructure/build.gradle:
```diff
@@ -1 +1,134 @@
+    processedTemplateFile = file("${buildDir}/template.json")
+    outputTemplateFile = file("${buildDir}/template.output.json")
+
+    stackOutputsFile = file("${buildDir}/${stackName}-${awsProfile}.outputs.json")
```
Again, I don't think we should be including the profile name in the file name. What's the advantage of doing that?

wjbuys replied:
See comment above about the purpose of the profile flag. This was merely the quickest fix, I'm open to any better suggestions.

narehayrapetyan commented on frontend-service/src/main/java/com/amazonaws/blox/frontend/LambdaHandler.java:
```diff
@@ -0,0 +1,65 @@
+
+/** Entrypoint for mapping incoming API GW requests to Lambda into Spring controllers */
+public class LambdaHandler implements RequestHandler<AwsProxyRequest, AwsProxyResponse> {
+  Logger log = Logger.getLogger(LambdaHandler.class);
```
Lombok has logger annotations we can use to skip this line everywhere

wjbuys replied:
Oh, neat! I didn't add a lombok dependency to the main project yet, so it's not available in this code at the moment. Will add.

narehayrapetyan commented on frontend-service/src/main/java/com/amazonaws/blox/frontend/models/Environment.java:
```diff
@@ -0,0 +1,27 @@
+public final class Environment {
+  private final String name;
+
+  public final String getName() {
```
getter and constructor can be replaced with @value

wjbuys replied:
Same as above, we didn't have lombok in the project when I started writing this, will update.

wjbuys commented:
I'll be working in your feedback on the small-level code stuff; do you have any concerns over the larger project structure/design so far? Want to sanity check that this is the shape we're aiming for, before we build too much on top of it.

narehayrapetyan commented:
looks fine overall

wjbuys commented:
Okay, as discussed in person, I've opened a separate issue for the testing stack naming/configuration stuff related to the AWS properties in the build script: #228

Commit:
Fix annoying compiler warning from transitive dependency

The swagger-maven-plugin package has a bad transitive dependency on an
old version of ANTLR that prints a warning on every build. This excludes
the problematic dependency, since it powers a feature of the plugin that
we're not using at all.

wjbuys commented:
Are there any specific changes you'd still like on this CR that aren't covered by #228?

narehayrapetyan commented on end-to-end-tests/src/test/java/com/amazonaws/blox/EnvironmentTest.java:
```diff
@@ -0,0 +1,52 @@
+import org.junit.Test;
+
+public class EnvironmentTest {
+  Blox client = Blox.builder().endpoint(endpoint()).build();
```
private?

wjbuys replied:
Must be public to be a runnable test, I believe.

wjbuys replied:
Ooooh, you mean the `client` var. Sorry, can't see exactly which line you commented on :/ Will fix.

narehayrapetyan commented on frontend-service-client/src/main/java/com/amazonaws/blox/AbstractBlox.java:
```diff
@@ -0,0 +1,31 @@
+/**
+
```
Why is this empty?

narehayrapetyan replied:
Oh is this generated code?

wjbuys replied:
Yup. I'll follow up and see if we can specify a license block or something in the API call that generates this, but I don't remember anything like that...
